### PR TITLE
(Not on a sandbox) Ticket #1501: Delete button for DomainApplication records

### LINKED ISSUE #1501: Allow users to delete domain requests they've started or withdrawn
### Issue description

To respect user's privacy, we should enable the ability for users to delete domain applications before they have submitted them or after they have withdrawn them.

### Acceptance criteria

- [ ] domain managers can delete domain applications before they have been submitted
- [ ] domain managers can delete domain applications that were withdrawn
- [ ] Implement design as defined in the [Figma mockup](https://www.figma.com/file/aJbuDLJNNikqQObTuCNKQa/GetGov_Designs_ECS-Truss?type=design&node-id=2955%3A12612&mode=design&t=Xf9KhN638FAIhaiC-1) Click the "present" button at the top right to see the prototype. 
- [ ] ~Number new domain requests that do not have a requested domain (see #1482)~ this was deemed too difficult, the creation date will be shown instead

### Additional context

_No response_

### Links to other issues

Design ticket: #735 
Related to: #1482

### PR DESCRIPTION
**Pull request by zandercymatics**
## Ticket

Resolves #1501

## Changes

- Added some styles in _tables.scss for positioning, increased width of card on the home page
- Modified home.html to include the delete button
- Added unit tests in test_views.py
- Added a new delete view in application.py for DomainApplication w/ a has_permission override to forbid certain domains from being deleted
- Modified index.py to add a context for the modal button and if there are any applications which can be deleted
- Added an abstract permission view for DomainApplication
- Adds some logic for displaying multiple unfinished applications

## Context for reviewers
This ticket adds a delete button for DomainApplications on the home page, and some logic where when you start multiple new domain applications, they will be numbered rather than having duplicate names. 

Regarding the delete button, it is only visible if the domain is in the state of started or withdrawn, otherwise it cannot be deleted. 

## Setup
**For designers**
Everything you need should already exists on my sandbox - one domainapplication in `STARTED` and one in `WITHDRAWN`. If this is not the case (i.e. there is not delete button for any of the DomainApplications), then you will need to navigate to /admin and add yourself as a submitter to a DomainApplication with either of those states.

Afterwards, just follow the workflow defined in the figma.

**For developers**
**IMPORTANT:** Since this PR adds a scss file and makes some of those changes, you will likely need to do collectstatic if that local cache is not refreshing for you. 

This PR should be fairly simple to review. When testing locally, all you need to do is boot up the application and you will have an assortment to test with.

Afterwards, start two domain applications, but do not finish either of them. Note that they will be numbered...

On sandboxes, you should already have this, but if you do not - grab a few applications and set yourself as a submitter. Then, delete them.

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [x] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [x] Checked that the design translated visually
- [x] Checked behavior
- [x] Checked different states (empty, one, some, error)
- [x] Checked for landmarks, page heading structure, and links
- [x] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [x] Checked keyboard navigability
- [x] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [x] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

## Screenshots
image

### CODE REVIEW

File: src/registrar/assets/sass/_theme/_usa-modal.scss
Comment (zandercymatics) [inline src/registrar/assets/sass/_theme/_usa-modal.scss:13]: This is the same color as the figma but its not consistent with the other modals in our app (consistency difference between that and the figma). Thoughts?

File: src/registrar/templates/home.html
Comment (zandercymatics) [inline src/registrar/templates/home.html:53]: Technically we don't need to add this, just up the width -- but this seems to work better with scaling. Its the most reliable solution I found for it so far, but I'd be open to other suggestions

File: src/registrar/tests/test_views.py
Comment (zandercymatics) [inline src/registrar/tests/test_views.py:159]: This can be re-implemented to exclude the for loop but it doesn't make sense to do here in this context. This is based off of other unit tests using subTest which test on status as well (FSM checks)

File: src/registrar/assets/sass/_theme/_usa-modal.scss
Comment (gabydisarli) [reply src/registrar/assets/sass/_theme/_usa-modal.scss:13]: @zandercymatics sometimes I forget to update the color when pulling in the USWDS base modal component because the color is so similar to our .gov palette. Go ahead and match the color we have for our styles on this and any modals going forward.

File: src/registrar/templates/home.html
Comment (gabydisarli) [inline src/registrar/templates/home.html:188]: @zandercymatics is this the language that'd show when the domain request has no domain name on it yet. Like in the case of a request name being "New domain request"?

Comment (zandercymatics) [reply src/registrar/templates/home.html:188]: @gabydisarli Correct! You can test this yourself by starting a new domain request, then leaving midway through. Rather than what it is now (`Are you sure you want to delete your domain request?`) Would you like this to conform to the previous pattern defined in the figma? I.e.  `Are you sure you want to delete New domain request`? Or something like this with corrected grammar?

**Comment by gabydisarli** [comment]
Some accessibility findings. I saw these all on Voiceover, I can walk through it with you if it's unclear.

- The delete button is called out as a button, where as the others are called out as a link. Not sure how we want to handle this but this is another area where I'm fuzzy in what is best practice, might be something else for @rachidatecs to consider. Should they be read out as buttons or links?
- For the other actions, Manage, Edit, VO reads out the corresponding request name, like "Edit <domainrequestname>". The Delete button does not. Can we add this in for delete, so it matches the others?
- When opening the modal with VO on, I'm read the entire contents of the modal twice.

File: src/registrar/templates/home.html
Comment (gabydisarli) [reply src/registrar/templates/home.html:188]: @zandercymatics yes let's follow that pattern. 

Are you sure you want to delete `New domain request #`?

Comment (gabydisarli) [reply src/registrar/templates/home.html:188]: @zandercymatics with the numbering of unnamed requests. Let's set the 1st one to be called "New domain request 1" instead of "New domain request, New domain request 2, ..."

File: src/registrar/assets/sass/_theme/_usa-modal.scss
Comment (gabydisarli) [reply src/registrar/assets/sass/_theme/_usa-modal.scss]: I double checked and I sometimes mistakenly use medium-50 when it should be dark-60. My bad!

Comment (zandercymatics) [reply src/registrar/assets/sass/_theme/_usa-modal.scss]: Not a problem!

**Review by gabydisarli — APPROVED** [review]
LGTM, thanks Zander!

**Review by dave-kennedy-ecs — COMMENTED** [review]
I pulled branch and tested locally.  Functions as expected.  Nice feature that I will make use of for sure!  Seems in line with how to perform this in django.
I don't have a comment on the css, it generally looks good to me, although I typically defer to others on the team with regards to css.

Non-blocking comment: I tested for contacts which are attached to the domain application.  When an application is deleted, contacts can / will become orphaned.  Is this the expected behavior?  Or do we want to delete them if they are not attached to any other entities in the system?

Comment: If I have a list of several New domain requests, and I delete the one labeled, New domain request 2, the page reloads and still shows New domain request 2 (since the domain requests get numbered at page load).  Is this the expected behavior?  Similarly, if you edit an existing request with a number on it, got through some screens, and save and continue, when you come back to the list and edit the existing request with that same number, you are not guaranteed to get the same application.  The numbering is confusing me.

**Comment by zandercymatics** [comment]
@dave-kennedy-ecs 
> When an application is deleted, contacts can / will become orphaned. Is this the expected behavior?

Great observation. I think this depends on the level of data integrity we want to maintain. If the contact is associated with a User we would not want to delete that, but otherwise you could make a solid argument for deleting this. This would be a good architectural discussion as it relates to the discussion we're having with the user/contact linkage

**Comment by zandercymatics** [comment]
>  The delete order is weird and names aren't consistent. Is this the expected behavior?  The numbering is confusing me.

Both of the issues you mentioned are a result of this information being generated on page load, rather than saving that change in the DB. This is generally unavoidable given the structure of those records.

To remedy this, a new DraftDomain object would need to be created each time a "New domain request" is started, which would mean the DB would be populated with numerous DraftDomains of the name "New domain request 1" (etc...). This makes me think that maybe we should add another field to DraftDomain, and some sort of mixin or something that would iterate this other field based on how many "None" exist

File: src/registrar/assets/sass/_theme/_usa-modal.scss
Comment (rachidatecs) [inline src/registrar/assets/sass/_theme/_usa-modal.scss:4]: Did you try the USWDS large modal `usa-modal--lg`? 

I think that could work. Related - if you ever actually need to extend a base CSS class with a variant, I like the -- convention (usa-modal--wider)

File: src/registrar/assets/sass/_theme/_tables.scss
Comment (rachidatecs) [inline src/registrar/assets/sass/_theme/_tables.scss:34]: Looks like this media query is missing a value. In any case, you should use the USWDS mixin (eg: `@include at-media('tablet')`)

File: src/registrar/assets/sass/_theme/_base.scss
Comment (rachidatecs) [inline src/registrar/assets/sass/_theme/_base.scss:129]: Missing ending blank line. Also there is a USWDS .text-no-underline class (https://designsystem.digital.gov/utilities/text-styles/). You should use it. If it doesn't work because of a cascade conflict, override with eg a.text-no-underline

File: src/registrar/templates/home.html
Comment (rachidatecs) [reply src/registrar/templates/home.html:53]: Why add this at all?

File: src/registrar/templates/home.html
Comment (rachidatecs) [inline src/registrar/templates/home.html:47]: Why is this needed?

**Review by rachidatecs — DISMISSED** [review]
See CSS and markup comments

File: src/registrar/assets/sass/_theme/_usa-modal.scss
Comment (rachidatecs) [reply src/registrar/assets/sass/_theme/_usa-modal.scss:4]: Looking at this again, the content should work fine in a default modal. We're using those for similarity sized content

Comment (zandercymatics) [reply src/registrar/assets/sass/_theme/_usa-modal.scss:4]: usa-modal--lg doesn't match the figma in terms of content-height and some other padding things. The reasoning for the implementation here is that most domain names tend to wrap their content, making it harder to read 
(stuff like this-plea-tea.gov becomes this-
plea-tea.gov). I think this should be kept for consistency with the design, but I am open to discussion on it

That said, I like the `usa-modal--wider` naming convention quite a lot

File: src/registrar/templates/home.html
Comment (zandercymatics) [reply src/registrar/templates/home.html:47]: By default, the "manage" button doesn't align with the manage button in the other table. This is especially apparent if the user doesn't have anything that they can delete, so there is just an empty column. ` has_deletable_applications` is a flag that switches between the default table display (as it was before this change w/o that column) and the new one. These two scenarios need different widths on the manage button, especially because it wasn't correctly aligned before this PR

File: src/registrar/assets/sass/_theme/_usa-modal.scss
Comment (therealslimhsiehdy) [inline src/registrar/assets/sass/_theme/_usa-modal.scss:6]: **Question:** I see that `!important` is used for overriding, and am curious what component you're overriding + reasoning for it? (Just trying to understand, not a "this is bad!" comment)

File: src/registrar/templates/home.html
Comment (therealslimhsiehdy) [inline src/registrar/templates/home.html:94]: **Question:** Is this just used to empty spacing?

File: src/registrar/views/index.py
Comment (therealslimhsiehdy) [inline src/registrar/views/index.py:21]: **Question:** Is it worth testing this scenario of "user exits without putting in a URL yet and want to make sure naming shows up correctly"? Don't know if it's worth (I also don't see in the Figma a sketch), but thought it'd be worth a mention

**Review by therealslimhsiehdy — COMMENTED** [review]
Screenshot 2024-01-10 at 12 59 48 PM
Screenshot 2024-01-10 at 2 18 39 PM

Left a few questions -- looks great in the mobile and desktop view (you can see draft domain naming of "New domain request 1" + delete button)

**Comment by rachidatecs** [comment]
> I pulled branch and tested locally. Functions as expected. Nice feature that I will make use of for sure! Seems in line with how to perform this in django. I don't have a comment on the css, it generally looks good to me, although I typically defer to others on the team with regards to css.
> 
> Non-blocking comment: I tested for contacts which are attached to the domain application. When an application is deleted, contacts can / will become orphaned. Is this the expected behavior? Or do we want to delete them if they are not attached to any other entities in the system?
> 
> Comment: If I have a list of several New domain requests, and I delete the one labeled, New domain request 2, the page reloads and still shows New domain request 2 (since the domain requests get numbered at page load). Is this the expected behavior? Similarly, if you edit an existing request with a number on it, got through some screens, and save and continue, when you come back to the list and edit the existing request with that same number, you are not guaranteed to get the same application. The numbering is confusing me.

+ 1 on numbering

On cleaning up orphan users: Dave and I did that on the other contacts formset delete. We should at the very least keep track of this side effect for a future ticket and / or for when we eventually refactor contacts.

File: src/registrar/assets/sass/_theme/_usa-modal.scss
Comment (zandercymatics) [reply src/registrar/assets/sass/_theme/_usa-modal.scss]: This is for the modal width just to increase how much wiggle room we get on that. This ended up being removed after a conversation with Rachid yesterday

File: src/registrar/views/index.py
Comment (zandercymatics) [reply src/registrar/views/index.py:21]: Yeah that's a good point. I am going to refactor how this bit works so I'll include that there

File: src/registrar/models/draft_domain.py
Comment (rachidatecs) [inline src/registrar/models/draft_domain.py:24]: I think we might me able to do without is_complete. Trick would be to refactor what we're passing to the modal and to give it a 'New domain request {{ application.created_at }}' - which may be tricky, but worth it.

File: src/registrar/templates/includes/modal.html
Comment (abroddrick) [inline src/registrar/templates/includes/modal.html:8]: same comment here about using a br instead of handling it with styling

File: src/registrar/views/index.py
Comment (abroddrick) [inline src/registrar/views/index.py:45]:
```suggestion
    applications = DomainApplication.objects.filter(creator=request.user).exclude(status=DomainApplication.ApplicationStatus.APPROVED
```
please change this before merging to use the contant value

**Review by abroddrick — APPROVED** [review]
See some minor comments before merging

File: src/registrar/templates/home.html
Comment (zandercymatics) [reply src/registrar/templates/home.html:144]: Great suggestion!

File: src/registrar/models/domain_application.py
Comment (zandercymatics) [reply src/registrar/models/domain_application.py:634]: Nope, nice find

File: src/registrar/templates/home.html
Comment (zandercymatics) [reply src/registrar/templates/home.html:50]: Will remove this as it is likely it won't be needed, rather it'd probably be better to just do sizing directly

**Review by abroddrick — CHANGES_REQUESTED** [review]
Sorry for approving before, I forgot to check that contacts were being deleted. That needs to happen here so we don't create orphaned contacts on our production environment.

**Review by gabydisarli — APPROVED** [review]
Looks great!

File: src/registrar/views/application.py
Comment (zandercymatics) [inline src/registrar/views/application.py:684]: Not the most elegant thing ever but it works. Let me know if this needs changing

File: src/registrar/views/application.py
Comment (zandercymatics) [inline src/registrar/views/application.py:664]: This is being done at the form level, as the ticket which handled the joins was done at this level too. From my understanding, we are trying to avoid modifying model-level behavior where possible, and this could be labelled as falling under that triage.

For a future ticket (probably during some sort of refactor or reconciliation), I would recommend some sort of helper, utility, class, etc that handles orphaned records more generically. This can be done either at the model or form level, but that on its own is another beast (metaprogramming) - but it can be done. I do not think this should be done now, though. Only if we have a need for it

File: src/registrar/views/application.py
Comment (abroddrick) [inline src/registrar/views/application.py:682]:
```suggestion
        # Pair each contact with its db related name for use in checking if it has joins
```

File: src/registrar/views/application.py
Comment (abroddrick) [reply src/registrar/views/application.py:684]: Agreed but makes sense here, just would update the comment to make it easier to read why this is happening

File: src/registrar/views/application.py
Comment (abroddrick) [inline src/registrar/views/application.py:667]: **change before merging**: add info on what this returns since it's not initially clear

File: src/registrar/views/application.py
Comment (abroddrick) [inline src/registrar/views/application.py:698]: **fix before merging**: this is only ever used for contacts, I would be more specific here this is for contacts, especially since you use the variable "contact" below. Also, why not just convert objects to a set and then back to a list? That will automatically remove duplicates in a list normally, though I have never tried with a query result so curious if you ran into an issue there.

File: src/registrar/tests/test_views.py
Comment (abroddrick) [inline src/registrar/tests/test_views.py:239]:
```suggestion
        # igorville is now deleted
        self.assertNotContains(response, "igorville.gov")
```

**Review by abroddrick — APPROVED** [review]
Looks good, thanks for the change. A couple minor suggestions and for sure remove the availability check change before merging. That really should be linked to the associated ticket in it's own PR

File: src/registrar/views/application.py
Comment (zandercymatics) [reply src/registrar/views/application.py:698]: The reason why I am not converting it to a set is because I actually want to get a list of duplicates. This is because duplicates let me detect an edge case:

A contact (which is not joined on any other field) is attached to both the authorizing_official field and the submitter field. When the delete is ran, it "looks" like this field is not an orphan (because it technically isn't) - but that isn't correct, because it will be after the delete.

This is basically a check to see if we run into that case, i.e. if the contact is shared among any of the fields on Application. If it is, there is a good chance that we hit this edge case - so we use this as a flag to determine if we should do a second join check or not

**Pull request merged**
